File: src/types/coCreation.ts
export const CO_CREATION_PROTOCOL_VERSION = 1 as const;

export type CoCreationStage =
  | 'story_seed'
  | 'creative_intent'
  | 'world_background'
  | 'rule_system'
  | 'protagonist'
  | 'core_conflict'
  | 'story_arc'
  | 'outline'
  | 'chapter_plan'
  | 'chapter_generation';

export type CoCreationStageStatus =
  | 'not_started'
  | 'in_progress'
  | 'minimum_complete'
  | 'complete'
  | 'skipped';

export type CoCreationIntent =
  | 'answer_current_question'
  | 'free_discussion'
  | 'modify_setting'
  | 'request_ai_completion'
  | 'generate_outline'
  | 'generate_chapter'
  | 'revise_existing_content'
  | 'accept_suggestion'
  | 'reject_suggestion'
  | 'undo_change'
  | 'navigate_to_page';

export type CoCreationFieldState =
  | 'user_confirmed'
  | 'ai_suggested'
  | 'ai_inferred'
  | 'temporary_assumption'
  | 'conflict'
  | 'blank';

export type CoCreationSourceType =
  | 'author_message'
  | 'formal_project_data'
  | 'adopted_chapter_text'
  | 'pending_draft'
  | 'ai_inference';

export interface CoCreationSourceReferenceV1 {
  sourceType: CoCreationSourceType;
  sourceId: string;
  excerpt?: string;
  contentHash?: string;
}

export interface CoCreationConflictV1 {
  code: string;
  severity: 'warning' | 'blocking';
  message: string;
  sourceReferences: CoCreationSourceReferenceV1[];
}

export interface CoCreationFieldTargetV1 {
  objectType:
    | 'story_seed'
    | 'creative_intent'
    | 'world_setting'
    | 'rule_system'
    | 'protagonist'
    | 'outline'
    | 'volume'
    | 'chapter';
  objectId?: string;
  fieldPath: string;
}

export interface CoCreationExtractedInformationV1 {
  target: CoCreationFieldTargetV1;
  value: unknown;
  fieldState: CoCreationFieldState;
  sourceReferences: CoCreationSourceReferenceV1[];
  confidence: number;
}

export interface CoCreationFieldSuggestionV1 {
  suggestionId: string;
  target: CoCreationFieldTargetV1;
  originalValue: unknown;
  suggestedValue: unknown;
  fieldState: Exclude<CoCreationFieldState, 'user_confirmed' | 'blank'>;
  sourceType: CoCreationSourceType;
  sourceReferences: CoCreationSourceReferenceV1[];
  confidence: number;
  conflicts: CoCreationConflictV1[];
  baseDataRevision: number;
  baseContextHash?: string;
  baseTargetVersion?: number;
  baseTargetHash?: string;
  decision: 'pending' | 'accepted_to_draft' | 'rejected';
  candidateHash: string;
  sourceMessageId?: string;
  sourceTaskId?: string;
  sourceArtifactId?: string;
  conflictsAcknowledged?: boolean;
  confirmedReplacement?: boolean;
  formalApplyPlanId?: string;
  formalAppliedAt?: string;
}

export interface CoCreationQuestionV1 {
  question: string;
  reason: string;
  targetFieldPaths: string[];
}

export interface CoCreationQuickReplyV1 {
  id: string;
  label: string;
  value: string;
}

export interface CoCreationStageCompletionV1 {
  stage: CoCreationStage;
  status: CoCreationStageStatus;
  completedRequiredFields: string[];
  missingRequiredFields: string[];
  percentage: number;
}

export interface CoCreationTurnOutputV1 {
  schemaVersion: typeof CO_CREATION_PROTOCOL_VERSION;
  naturalLanguageReply: string;
  intent: CoCreationIntent;
  currentStage: CoCreationStage;
  extractedInformation: CoCreationExtractedInformationV1[];
  pendingConfirmations: string[];
  nextHighValueQuestion?: CoCreationQuestionV1;
  quickReplies: CoCreationQuickReplyV1[];
  changeSuggestions: CoCreationFieldSuggestionV1[];
  stageCompletion: CoCreationStageCompletionV1;
  dataRevision: number;
}

export interface CoCreationStageProgress {
  stage: CoCreationStage;
  status: CoCreationStageStatus;
  percentage: number;
  missingRequiredFields: string[];
}

export interface CoCreationObjectContext {
  novelId: string;
  volumeId?: string;
  chapterId?: string;
  objectType?: string;
  objectId?: string;
  selectedText?: string;
  selectedTextHash?: string;
}

export interface CoCreationTurnContextV1 {
  currentStage: CoCreationStage;
  canonicalDataHash: string;
  dataRevision: number;
}

export interface CoCreationSession {
  sessionId: string;
  novelId: string;
  title: string;
  status: 'active' | 'archived';
  currentStage: CoCreationStage;
  stageProgress: CoCreationStageProgress[];
  objectContext: CoCreationObjectContext;
  summary?: string;
  summaryHash?: string;
  activeDraftRevisionId?: string;
  activeArtifactId?: string;
  dataRevision: number;
  dataHash: string;
  createdAt: string;
  updatedAt: string;
  archivedAt?: string;
}

export interface CoCreationMessage {
  messageId: string;
  sessionId: string;
  sequenceNo: number;
  role: 'user' | 'assistant';
  status: 'pending' | 'completed' | 'failed' | 'cancelled';
  content: string;
  contentHash: string;
  contentLength: number;
  replyToMessageId?: string;
  sourceTaskId?: string;
  sourceArtifactId?: string;
  turnContext?: CoCreationTurnContextV1;
  structuredPayload?: unknown;
  operationId: string;
  requestHash: string;
  createdAt: string;
  completedAt?: string;
}

export interface CoCreationDraftRevision {
  draftRevisionId: string;
  sessionId: string;
  stage: CoCreationStage;
  revisionNo: number;
  parentRevisionId?: string;
  schemaVersion: number;
  payload: Record<string, unknown>;
  contentHash: string;
  origin: 'author_edit' | 'assistant_proposal_accepted' | 'assistant_turn';
  sourceMessageId?: string;
  sourceTaskId?: string;
  sourceArtifactId?: string;
  operationId: string;
  requestHash: string;
  createdAt: string;
}

export interface CoCreationWorkspaceSnapshot {
  session: CoCreationSession;
  messages: CoCreationMessage[];
  draftRevisions: CoCreationDraftRevision[];
  activeDraft?: CoCreationDraftRevision;
  pendingTurn?: CoCreationMessage;
}

export interface PersistedCoCreationSessionV1 {
  sessionId: string;
  novelId: string;
  workspaceType: string;
  status: 'active' | 'archived';
  revision: number;
  stateHash: string;
  createdAt: string;
  updatedAt: string;
  archivedAt?: string;
}

export interface PersistedCoCreationMessageV1 {
  messageId: string;
  sessionId: string;
  turnId: string;
  sequenceNo: number;
  role: 'user' | 'assistant';
  status: 'submitted' | 'running' | 'completed' | 'failed' | 'cancelled';
  content: string;
  contentHash: string;
  contentLength: number;
  replyToMessageId?: string;
  taskId?: string;
  artifactId?: string;
  turnContext?: CoCreationTurnContextV1;
  error?: unknown;
  createdAt: string;
  completedAt?: string;
}

export interface PersistedCoCreationDraftRevisionV1 {
  draftRevisionId: string;
  sessionId: string;
  stageKey: CoCreationStage;
  revisionNo: number;
  parentRevisionId?: string;
  schemaVersion: number;
  payload: Record<string, unknown>;
  contentHash: string;
  origin: CoCreationDraftRevision['origin'];
  sourceMessageId?: string;
  sourceTaskId?: string;
  sourceArtifactId?: string;
  createdAt: string;
}

export interface PersistedCoCreationWorkspaceV1 {
  schemaVersion: number;
  session: PersistedCoCreationSessionV1;
  messages: PersistedCoCreationMessageV1[];
  draftRevisions: PersistedCoCreationDraftRevisionV1[];
}

export interface CoCreationMutationReceiptV1 {
  sessionId: string;
  operationId: string;
  operationType: string;
  revision: number;
  stateHash: string;
  messageId?: string;
  draftRevisionId?: string;
  idempotentReplay: boolean;
}

File: src-tauri/src/domain/co_creation.rs
use serde::{Deserialize, Serialize};
use serde_json::Value;

pub const CO_CREATION_SCHEMA_VERSION: i64 = 1;
pub const CO_CREATION_WORKSPACE_TYPE: &str = "ai_co_creation";

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct CoCreationTurnContextV1 {
    pub current_stage: String,
    pub canonical_data_hash: String,
    pub data_revision: i64,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct CoCreationSessionV1 {
    pub session_id: String,
    pub novel_id: String,
    pub workspace_type: String,
    pub status: String,
    pub revision: i64,
    pub state_hash: String,
    pub created_at: String,
    pub updated_at: String,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub archived_at: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct CoCreationMessageV1 {
    pub message_id: String,
    pub session_id: String,
    pub turn_id: String,
    pub sequence_no: i64,
    pub role: String,
    pub status: String,
    pub content: String,
    pub content_hash: String,
    pub content_length: i64,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub reply_to_message_id: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub task_id: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub artifact_id: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub turn_context: Option<CoCreationTurnContextV1>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub error: Option<Value>,
    pub created_at: String,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub completed_at: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct CoCreationDraftRevisionV1 {
    pub draft_revision_id: String,
    pub session_id: String,
    pub stage_key: String,
    pub revision_no: i64,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub parent_revision_id: Option<String>,
    pub schema_version: i64,
    pub payload: Value,
    pub content_hash: String,
    pub origin: String,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub source_message_id: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub source_task_id: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub source_artifact_id: Option<String>,
    pub created_at: String,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct CoCreationWorkspaceV1 {
    pub schema_version: i64,
    pub session: CoCreationSessionV1,
    pub messages: Vec<CoCreationMessageV1>,
    pub draft_revisions: Vec<CoCreationDraftRevisionV1>,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq)]
#[serde(rename_all = "camelCase")]
pub struct OpenCoCreationWorkspaceResultV1 {
    pub created: bool,
    pub workspace: CoCreationWorkspaceV1,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct CoCreationMutationReceiptV1 {
    pub session_id: String,
    pub operation_id: String,
    pub operation_type: String,
    pub revision: i64,
    pub state_hash: String,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub message_id: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub draft_revision_id: Option<String>,
    pub idempotent_replay: bool,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct OpenCoCreationWorkspaceInput {
    pub novel_id: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct ReadCoCreationWorkspaceInput {
    pub novel_id: String,
    pub session_id: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct RecoverCoCreationTurnTaskInput {
    pub novel_id: String,
    pub session_id: String,
    pub user_message_id: String,
}

#[derive(Debug, Clone, Serialize, Deserialize, PartialEq, Eq)]
#[serde(rename_all = "camelCase")]
pub struct RecoveredCoCreationTurnTaskV1 {
    pub task_id: String,
    pub current_stage: String,
    pub canonical_data_hash: String,
    pub data_revision: i64,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct AppendCoCreationUserMessageInput {
    pub novel_id: String,
    pub session_id: String,
    pub expected_revision: i64,
    pub expected_state_hash: String,
    pub operation_id: String,
    #[serde(default)]
    pub request_hash: Option<String>,
    pub content: String,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct BindCoCreationTurnTaskInput {
    pub novel_id: String,
    pub session_id: String,
    pub user_message_id: String,
    pub task_id: String,
    pub expected_revision: i64,
    pub expected_state_hash: String,
    pub operation_id: String,
    #[serde(default)]
    pub request_hash: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct CompleteCoCreationTurnInput {
    pub novel_id: String,
    pub session_id: String,
    pub user_message_id: String,
    pub task_id: String,
    pub artifact_id: String,
    pub expected_revision: i64,
    pub expected_state_hash: String,
    pub operation_id: String,
    #[serde(default)]
    pub request_hash: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct FailCoCreationTurnInput {
    pub novel_id: String,
    pub session_id: String,
    pub user_message_id: String,
    pub task_id: String,
    pub error_code: String,
    pub error_message: String,
    pub expected_revision: i64,
    pub expected_state_hash: String,
    pub operation_id: String,
    #[serde(default)]
    pub request_hash: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(rename_all = "camelCase", deny_unknown_fields)]
pub struct SaveCoCreationDraftRevisionInput {
    pub novel_id: String,
    pub session_id: String,
    pub stage_key: String,
    pub schema_version: i64,
    pub payload: Value,
    pub origin: String,
    #[serde(default)]
    pub source_message_id: Option<String>,
    #[serde(default)]
    pub source_task_id: Option<String>,
    #[serde(default)]
    pub source_artifact_id: Option<String>,
    pub expected_draft_revision: i64,
    #[serde(default)]
    pub expected_draft_content_hash: Option<String>,
    pub expected_revision: i64,
    pub expected_state_hash: String,
    pub operation_id: String,
    #[serde(default)]
    pub request_hash: Option<String>,
}
